crates/hir-ty/src/diagnostics/expr.rs

```diff
@@ -149,17 +150,18 @@ impl ExprValidator {
                 None => return,
             };
 
-            if filter_map_next_checker
-                .get_or_insert_with(|| {
-                    FilterMapNextChecker::new(&self.owner.resolver(db.upcast()), db)
-                })
-                .check(call_id, receiver, &callee)
-                .is_some()
-            {
+            let checker = filter_map_next_checker.get_or_insert_with(|| {
+                FilterMapNextChecker::new(&self.owner.resolver(db.upcast()), db)
+            });
+
+            if checker.check(call_id, receiver, &callee).is_some() {
                 self.diagnostics.push(BodyValidationDiagnostic::ReplaceFilterMapNextWithFindMap {
                     method_call_expr: call_id,
                 });
             }
+
+            let receiver_ty = self.infer[*receiver].strip_references().clone();
```

Comment:
Why are we stripping it here? If this is due to coercions and autoref then we should check the adjustments:
```rs
    let coerced = self.infer
            .expr_adjustments
            .get(receiver)
            .and_then(|adjusts| adjusts.last().map(|adjust| adjust.target.clone()));
```

Reply:
Ah nvm you want to get the dyn trait, then we should do the stripping where we check for the dyn trait, that seems more proper.